### PR TITLE
transport: discard droppable msg when channel full

### PR DESCRIPTION
**Pull request by BusyJay**
ref: https://github.com/pingcap/tikv/issues/1028

@ngaut @siddontang @disksing @zhangjinpeng1987 PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
why not using send + try_send?

File: src/util/transport.rs [hunk 1]
```diff
@@ -41,25 +55,28 @@ impl<T: Debug> SendCh<T> {
...
-            t = match e {
-                NotifyError::Full(m) => m,
-                _ => return Err(box_err!("{:?}", e)),
+    pub fn send(&self, t: T) -> Result<(), Error> {
+        self.try_send(t, 1)
+    }
+
+    pub fn try_send(&self, mut t: T, mut try_times: usize) -> Result<(), Error> {
+        loop {
+            t = match self.ch.send(t) {
```
Comment (ngaut) [inline src/util/transport.rs]: Any tests to cover the two branches?

Comment (BusyJay) [inline src/util/transport.rs]: Yes, see L152.

**Comment by BusyJay** [comment]
PTAL

**Comment by siddontang** [comment]
LGTM

PTAL @hhkbp2

**Comment by BusyJay** [comment]
PTAL

File: src/storage/txn/scheduler.rs [hunk 2]
```diff
@@ -41,7 +41,7 @@ use storage::mvcc::{MvccTxn, MvccReader, Error as MvccError};
 use storage::{Key, Value, KvPair};
 use std::collections::HashMap;
 use mio::{self, EventLoop};
-use util::transport::SendCh;
+use util::transport::{SendCh, MAX_SEND_RETRY_CNT};
```
Comment (hhkbp2) [inline src/storage/txn/scheduler.rs]: MAX_SEND_RETRY_CNT is a global constant and used everywhere when calling `try_send()`. Maybe it's good to  make retry count a attribute of this channel, and ignore it when calling `try_send()`.

File: src/util/transport.rs [hunk 3]
```diff
@@ -41,25 +55,33 @@ impl<T: Debug> SendCh<T> {
...
-            t = match e {
-                NotifyError::Full(m) => m,
-                _ => return Err(box_err!("{:?}", e)),
+    pub fn send(&self, t: T) -> Result<(), Error> {
+        self.send_with_try_times(t, 1)
+    }
+
+    /// Try send t with default try times.
+    pub fn try_send(&self, t: T) -> Result<(), Error> {
+        self.send_with_try_times(t, MAX_SEND_RETRY_CNT)
```
Comment (siddontang) [inline src/util/transport.rs]: I prefer send using retry count but try_send using only 1, like channel send and try_send.

**Comment by hhkbp2** [comment]
LGTM

File: src/util/transport.rs [hunk 4]
```diff
@@ -23,7 +23,11 @@ const MAX_SEND_RETRY_CNT: usize = 5;
 quick_error! {
     #[derive(Debug)]
     pub enum Error {
-        Other(err: Box<error::Error + Sync + Send>) {
+        Discard(reason: String) {
+            description("message is discard")
```
Comment (disksing) [inline src/util/transport.rs]: message is **discarded**

File: src/util/transport.rs [hunk 5]
```diff
@@ -32,6 +36,16 @@ quick_error! {
     }
 }
 
+impl<T: Debug> From<NotifyError<T>> for Error {
+    fn from(e: NotifyError<T>) -> Error {
+        match e {
+            // ALLERT!! make cause sensitive data leak.
```
Comment (disksing) [inline src/util/transport.rs]: make -> may?

**Comment by disksing** [comment]
LGTM.

**Comment by BusyJay** [comment]
@siddontang PTAL

**Comment by siddontang** [comment]
LGTM

**Pull request merged**